File: 06/CCC '06 J4 - It's tough being a teen!.py
c = [[] for x in range(8)]

c[1] = [2]
c[4] = [1, 3]
c[5] = [3]
c[7] = [1]

while 1:
    a, b = int(input()), int(input())

    if a == 0 and b == 0: break 

    c[b].append(a)

for x in range(1, 8):
    c[x].sort()

ans = []

for _ in range(7):
    for x in range(1, 8):
        if str(x) not in ans:
            if not c[x]:
                ans.append(str(x))
                for y in range(1, 8):
                    if x in c[y]:
                        c[y].remove(x)
                break

if len(ans) == 7:
    print(" ".join(ans))
else:
    print("Cannot complete these tasks. Going to bed.")

File: 06/CCC '06 J4 - It's tough being a teen!.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> graph[8];
int indeg[8]; 

void topsort() {
    vector<int> q;
    int cnt = 0; 
    int order[8];

    for (int x = 1; x <= 7; x++) {
        if (indeg[x] == 0) q.push_back(x);
    }

    while (!q.empty()) {
        sort(q.begin(), q.end(), greater<int>());
        int v = q.back();
        q.pop_back();
        cnt ++;
        order[cnt] = v; 

        for (int x: graph[v]) {
            indeg[x] --;
            if (indeg[x] == 0) {
                q.push_back(x);
            } 
        }
    }
    if (cnt != 7) {
        cout << "Cannot complete these tasks. Going to bed.";
    }
    else {
        for (int x = 1; x <= 7; x++) cout << order[x] << " ";
    }
}

int main() {
    cin.sync_with_stdio(0); cin.tie(0);    

    indeg[1] = 1; 
    indeg[2] = 0;
    indeg[3] = 0; 
    indeg[4] = 2; 
    indeg[5] = 1;
    indeg[6] = 0; 
    indeg[7] = 1; 

    graph[1] = {4, 7};
    graph[2] = {1};
    graph[3] = {4, 5};

    while (true) {
        int a, b;
        cin >> a >> b;
        if (a == 0 && b == 0) break; 

        graph[a].push_back(b);
        indeg[b] ++;
    }

    topsort();
}